### DISCUSSION TITLE
On pressure nullspaces

### DISCUSSION THREAD

**Discussion by sghelichkhani** [g-adopt/g-adopt#67] [General]
### Nullspace
The **nullspace** (or kernel) of a matrix $M$ is defined as the set of all vectors $\mathbf{v}$ such that: $M\mathbf{v} = \mathbf{0}$
This includes all vectors that are mapped to the zero vector by the linear transformation represented by $M$. The dimension of this space, known as the nullity of $M$, indicates the number of linearly independent solutions to the equation $M\mathbf{v} = \mathbf{0}$.

### Left Nullspace (Transpose Nullspace)
The **left nullspace** (or transpose nullspace) of a matrix $M$ is the nullspace of $M^T$, consisting of all vectors $\mathbf{w}$ that satisfy: $M^T\mathbf{w} = \mathbf{0}$
Vectors in the left nullspace are orthogonal to the columns of $M$, meaning each vector $\mathbf{w}$ is orthogonal to every column vector of $M$.

## Approximations and Equations
In G-ADOPT, and in general for geodynamic applications, we have three approximations for the Stokes equations. [All the bar'ed variables (e. g., $\bar{\rho}$) are only radially varying].

| Approximation | Equations |
|---------------|-----------|
| **(Extended) Boussinesq Approximation** | $$\nabla \cdot u = 0$$ <br> $$\nabla \cdot \mu [\nabla u + (\nabla u)^T] - \nabla p = \hat{e}_{r} Ra T$$ |
| **Truncated Anelastic Liquid Approximation (TALA)** | $$\nabla \cdot \left(\bar{\rho} u\right) = 0$$ $$\nabla \cdot \mu [\nabla u + (\nabla u)^T - (2/3) \nabla \cdot u I] - \nabla p = \bar{g}\bar{\rho}\bar{\alpha}\hat{e}_{r} Ra T$$ |
| **Anelastic Liquid Approximation (ALA)** |$$\nabla \cdot \left(\bar{\rho} u\right) = 0$$ $$\nabla \cdot mu [\nabla u + (\nabla u)^T - (2/3) \nabla \cdot u I] - \nabla p + \bar{g} \bar{\rho} \bar{\chi} (D / \gamma_r)(\bar{C_v} / \bar{C_v}) p = \bar{g}\bar{\rho}\bar{\alpha} \hat{e}_{r} Ra T$$ |

## Discrete Form and Pressure Nullspaces
In its simplest form of the discrete form, which is for the Boussinesq approximation, we have the notation:

$$
\left[\begin{array}{c}
M 
\end{array}\right] \left[\begin{array}{c}
u \\
p \\
\end{array}\right] = \left[\begin{array}{cc}
A & B^T \\
B & 0 \\
\end{array}\right] \left[\begin{array}{c}
u \\
p \\
\end{array}\right] = \left[\begin{array}{c}
f \\
0 \\
\end{array}\right]
$$

where $Au$ is the block associated with the strain-rate term, $B^T p$ is the gradient of pressure, and $B u$ is the continuity equation. When extending to TALA (see above), $M$ is not symmetric anymore, and it changes to,

$$
\left[\begin{array}{cc}
A^{\prime} & B^T \\
B_{\rho} & 0 \\
\end{array}\right]
$$

with $B_{\rho}$ being the term associated with the mass conservation equation $\nabla \cdot \left(\bar{\rho} u\right)$. Furthermore, extending to ALA, we will have

$$
\left[\begin{array}{cc}
A^{\prime} & B^T + B_p \\
B_{\rho} & 0 \\
\end{array}\right]
$$

with $B_p p$ associated with $\bar{g} \bar{\rho} \bar{\chi} (D/ \gamma_r)( \bar{C_p}/\bar{C_v}) p$. Considering all the above, we will have the following for (right) nullspaces and transpose (left) nullspaces

| Approximation | Discrete Form | (Right) Nullspace | Transpose (Left) Nullspace |
|---------------|---------------|--------------------|--------------------|
| **Boussinesq Approximation** | $`\left[\begin{array}{cc} A & B^T \\ B & 0 \end{array}\right] \left[\begin{array}{c}u \\ p\end{array}\right] = \left[\begin{array}{c}f \\ 0\end{array}\right]`$ | Solution to $B^T v = 0$, which in strong form is $\nabla p = 0$. or `nullspace=Constant`|   Solution to $B^T w = 0$, which in strong form is $\nabla p = 0$. or `transpose_nullspace=Constant`.
| **Truncated Anelastic Liquid Approximation (TALA)** | $`\left[\begin{array}{cc} A' & B^T \\ B_{\rho} & 0 \end{array}\right] \left[\begin{array}{c}u \\ p\end{array}\right] = \left[\begin{array}{c}f \\ 0\end{array}\right]`$ | *Like above* Solution to $B^T v = 0$, which in strong form is $\nabla p = 0$. or `nullspace=Constant` |Solution to $\left[B_p\right]^T \left[w\right] = 0$ , which in strong form is $` - \int \nabla p \cdot \bar{\rho} u = 0`$ which analytically is $$p = C$$ and hence `transpose_nullspace=Constant`.
| **Anelastic Liquid Approximation (ALA)** | $`\left[\begin{array}{cc} A' & B^T + \bar{B}_p \\ B_{\rho} & 0 \end{array}\right] \left[\begin{array}{c}u \\ p\end{array}\right] = \left[\begin{array}{c}f \\ 0\end{array}\right]`$ | Solution to $\left[B^{T}+ \bar{B}_p\right] v = 0$. which in strong form is the solution to  $\nabla p - \bar{g}\bar{\rho} \bar{\chi} (D / \gamma_r) (\bar{C_p} / \bar{C_v}) p =0$ or analytically $$p = C exp(\int_r \bar{g}\bar{\rho} \bar{\chi} (D / \gamma_r) (\bar{C_p} / \bar{C_v})  dr)$$ | *Like above* Solution to $\left[B_p\right]^T \left[w\right] = 0$, which in strong form is $` - \int \nabla p \cdot \bar{\rho} u = 0`$  which analytically is $$p = C$$ and hence `transpose_nullspace=Constant`.

**Comment by Patol75** [comment]
I tried to make the mathematical expressions in the final table render, but no luck. :/
